File: sdk/py/hop_top_kit/scope.py
"""hop_top_kit.scope — filesystem path policy guardrails.

Python port of ``hop.top/kit/go/core/scope``. Mirrors the Go API surface
(``new``, ``default``, ``allow``, ``deny``, ``check``, ``enforce``, ``set_mode``,
``set_default``, ``snapshot``, ``secret_paths``) using snake_case naming.

The default deny list is loaded from ``contracts/parity/scope-defaults.json``
(the cross-language source of truth). A local copy is bundled into the
package as ``scope-defaults.json``; ``test_scope.py`` asserts it matches the
canonical contract file.

Glob matching uses `pathspec <https://pypi.org/project/pathspec/>`_ in
GitWildMatch mode, which supports ``**`` recursive matching as needed by
the shared pattern set.

Example::

    from hop_top_kit.scope import default, Op, ErrDenied

    try:
        default().enforce("/Users/me/.ssh/id_rsa", Op.READ)
    except ErrDenied:
        ...  # path denied
"""

from __future__ import annotations

import json
import logging
import os
import sys
from collections.abc import Callable
from dataclasses import dataclass, field
from enum import IntFlag, StrEnum
from pathlib import Path
from typing import Optional

from pathspec import GitIgnoreSpec

__all__ = [
    "Decision",
    "ErrDenied",
    "Mode",
    "Op",
    "Policy",
    "default",
    "new",
    "secret_paths",
    "set_default",
]

_LOG = logging.getLogger(__name__)


# ─── Enums ───────────────────────────────────────────────────────────────────


class Mode(StrEnum):
    """Mode controls how :meth:`Policy.enforce` reacts to a Denied decision."""

    STRICT = "strict"
    """Strict denies → raises :class:`ErrDenied`. The default."""
    WARN = "warn"
    """Warn denies → logs at WARNING level and returns."""
    PROMPT = "prompt"
    """Prompt denies → invokes the prompt callback."""


class Op(IntFlag):
    """Op is a bitset of filesystem operations."""

    READ = 1 << 0
    WRITE = 1 << 1
    EXEC = 1 << 2


class Decision(StrEnum):
    """Decision is the outcome of :meth:`Policy.check`."""

    UNKNOWN = "unknown"
    ALLOWED = "allowed"
    DENIED = "denied"


# ─── Errors ──────────────────────────────────────────────────────────────────


class ErrDenied(Exception):  # noqa: N818  (mirrors Go ErrDenied — kept for cross-language API parity)
    """Raised by :meth:`Policy.enforce` when a path is denied."""

    def __init__(self, path: str, op: Op) -> None:
        super().__init__(f"scope: path denied: {path} (op={_op_string(op)})")
        self.path = path
        self.op = op


# ─── Rule + callback ─────────────────────────────────────────────────────────


@dataclass
class Rule:
    """A single rule: glob patterns + ops bitset + allow/deny verdict."""

    patterns: list[str]
    ops: Op
    allow: bool


PromptFunc = Callable[[str, Op], bool]
"""Prompt callback. Return ``True`` to allow this single call."""


# ─── Defaults JSON ───────────────────────────────────────────────────────────


_DEFAULTS_PATH = Path(__file__).parent / "scope-defaults.json"
_DEFAULTS_RAW = json.loads(_DEFAULTS_PATH.read_text(encoding="utf-8"))


def _platform_key() -> str:
    """Map ``sys.platform`` to the keys used in scope-defaults.json."""
    if sys.platform == "darwin":
        return "darwin"
    if sys.platform == "win32":
        return "windows"
    return "linux"


def _expand_windows_env(p: str) -> str:
    """Expand ``%APPDATA%``, ``%LOCALAPPDATA%``, ``%USERPROFILE%`` from env.

    On non-Windows hosts, the macros are left in place when the env var is
    unset; the pattern simply won't match anything on those hosts.
    """
    for key in ("APPDATA", "LOCALAPPDATA", "USERPROFILE"):
        token = f"%{key}%"
        if token not in p:
            continue
        v = os.environ.get(key, "")
        if not v:
            continue
        p = p.replace(token, v)
    return p


def secret_paths() -> list[str]:
    """Return the default deny pattern set: common + platform-specific.

    Patterns are ready to feed into :meth:`Policy.deny`. ``~`` and Windows
    env macros are resolved at match time (``~`` is left intact here so the
    set is host-portable; ``%FOO%`` is resolved against the current env).
    """
    common = _DEFAULTS_RAW["deny"].get("common", [])
    platform_specific = _DEFAULTS_RAW["deny"].get(_platform_key(), [])
    return [_expand_windows_env(p) for p in (*common, *platform_specific)]


# ─── Policy ──────────────────────────────────────────────────────────────────


@dataclass
class Policy:
    """Holds an ordered list of rules, a Mode, and an optional prompt callback.

    Construct via :func:`new` or :func:`default`.
    """

    _rules: list[Rule] = field(default_factory=list)
    _mode: Mode = Mode.STRICT
    _prompt: Optional[PromptFunc] = None

    # Mutators (return self for chaining) ────────────────────────────────────

    def allow(self, *patterns: str) -> Policy:
        """Register an allow rule covering Read|Write|Exec."""
        return self.allow_op(Op.READ | Op.WRITE | Op.EXEC, *patterns)

    def allow_op(self, op: Op, *patterns: str) -> Policy:
        """Register an allow rule for the given operations."""
        if not patterns:
            return self
        self._rules.append(Rule(patterns=list(patterns), ops=op, allow=True))
        return self

    def deny(self, *patterns: str) -> Policy:
        """Register a deny rule covering Read|Write|Exec."""
        return self.deny_op(Op.READ | Op.WRITE | Op.EXEC, *patterns)

    def deny_op(self, op: Op, *patterns: str) -> Policy:
        """Register a deny rule for the given operations."""
        if not patterns:
            return self
        self._rules.append(Rule(patterns=list(patterns), ops=op, allow=False))
        return self

    def set_mode(self, m: Mode) -> Policy:
        """Set the enforcement mode."""
        self._mode = m
        return self

    def get_mode(self) -> Mode:
        """Return the current enforcement mode."""
        return self._mode

    def set_prompt_func(self, fn: PromptFunc) -> Policy:
        """Set the prompt callback used in :class:`Mode.PROMPT`."""
        self._prompt = fn
        return self

    def get_rules(self) -> list[Rule]:
        """Defensive copy of the rules in registration order."""
        return [Rule(patterns=list(r.patterns), ops=r.ops, allow=r.allow) for r in self._rules]

    # Decision API ────────────────────────────────────────────────────────────

    def check(self, p: str, op: Op) -> Decision:
        """Evaluate (path, op). Pure — no prompt, no mutation.

        Resolves symlinks before matching to defeat symlink escapes; on
        FileNotFoundError, the cleaned path is used so "intent to write"
        still matches deny rules.
        """
        resolved = _resolve_path(p)
        saw_allow = False
        for r in self._rules:
            if (r.ops & op) == 0:
                continue
            if not _match_any(r.patterns, resolved):
                continue
            if not r.allow:
                return Decision.DENIED
            saw_allow = True
        return Decision.ALLOWED if saw_allow else Decision.UNKNOWN

    def enforce(self, p: str, op: Op) -> None:
        """Call :meth:`check` and translate the decision per current Mode.

        - ``Allowed`` → return.
        - ``Unknown`` + non-Strict → return.
        - ``Strict`` + (Denied | Unknown) → raise :class:`ErrDenied`.
        - ``Warn``   + Denied → log + return.
        - ``Prompt`` + Denied → invoke callback; raise on False.
        """
        dec = self.check(p, op)
        if dec is Decision.ALLOWED:
            return
        if dec is Decision.UNKNOWN and self._mode is not Mode.STRICT:
            return
        self._handle_deny(p, op)

    def _handle_deny(self, p: str, op: Op) -> None:
        if self._mode is Mode.WARN:
            _LOG.warning("scope: path denied (warn mode, allowing): %s op=%s", p, _op_string(op))
            return
        if self._mode is Mode.PROMPT:
            if self._prompt is not None and self._prompt(p, op):
                return
            raise ErrDenied(p, op)
        # Strict (default)
        raise ErrDenied(p, op)

    def snapshot(self) -> Policy:
        """Deep copy. Useful for per-test isolation paired with :func:`set_default`."""
        cp = Policy(_mode=self._mode, _prompt=self._prompt)
        cp._rules = [Rule(patterns=list(r.patterns), ops=r.ops, allow=r.allow) for r in self._rules]
        return cp


# ─── Singleton + factory ─────────────────────────────────────────────────────


_default_policy: Optional[Policy] = None


def new() -> Policy:
    """Create an empty :class:`Policy` in :class:`Mode.STRICT`."""
    return Policy()


def default() -> Policy:
    """Return the package-level singleton policy.

    Lazy-initialised on first call. Pre-seeded with :func:`secret_paths` as
    a deny-all-ops rule.
    """
    global _default_policy
    if _default_policy is None:
        _default_policy = new().deny(*secret_paths())
    return _default_policy


def set_default(p: Policy) -> Callable[[], None]:
    """Swap the singleton with ``p``. Returns a restore function.

    Intended for tests::

        restore = set_default(new().allow("~/tmp/**"))
        try:
            ...
        finally:
            restore()
    """
    global _default_policy
    prev = default()
    _default_policy = p

    def restore() -> None:
        global _default_policy
        _default_policy = prev

    return restore


# ─── Path resolution + match ─────────────────────────────────────────────────


def _expand_home(s: str) -> str:
    """Expand a leading ``~`` to the user home dir (resolving symlinks)."""
    if s == "~" or s.startswith("~/"):
        try:
            home = os.path.realpath(os.path.expanduser("~"))
        except OSError:
            home = os.path.expanduser("~")
        return home if s == "~" else os.path.join(home, s[2:])
    return s


def _resolve_path(s: str) -> str:
    """Resolve ``s``: expand ``~``, normalise, follow symlinks.

    On ENOENT walks up to the deepest existing ancestor and re-attaches the
    missing tail so deny rules match by intent.
    """
    expanded = _expand_home(s)
    cleaned = os.path.normpath(expanded)
    try:
        return os.path.realpath(cleaned, strict=True)
    except (FileNotFoundError, OSError):
        pass
    # Walk up.
    head, tail = os.path.split(cleaned)
    head = head.rstrip(os.sep)
    while head and head != os.sep:
        try:
            resolved = os.path.realpath(head, strict=True)
            return os.path.join(resolved, tail)
        except (FileNotFoundError, OSError):
            pass
        head, more = os.path.split(head)
        head = head.rstrip(os.sep)
        tail = os.path.join(more, tail)
    return cleaned


def _match_any(patterns: list[str], abs_path: str) -> bool:
    """Return True if any pattern matches abs_path (already resolved)."""
    for p in patterns:
        expanded = _expand_home(p)
        canon = _canonicalise_pattern(os.path.normpath(expanded))
        if _gitwildmatch(canon, abs_path):
            return True
    return False


def _gitwildmatch(pattern: str, path_str: str) -> bool:
    """Match ``path_str`` against a single gitignore-style ``pattern``."""
    # GitIgnoreSpec expects gitignore-relative semantics: leading "/" anchors
    # to root. To match absolute paths verbatim we strip the leading "/"
    # from both pattern and path before testing.
    p = pattern.lstrip(os.sep)
    a = path_str.lstrip(os.sep)
    spec = GitIgnoreSpec.from_lines([p])
    return bool(spec.match_file(a))


_GLOB_META = set("*?[]{}")


def _canonicalise_pattern(pat: str) -> str:
    """Resolve the leading literal directory prefix of ``pat`` through realpath.

    Mirrors the Go port: ensures ``/tmp/**`` keeps matching after the input
    path is canonicalised (e.g. ``/tmp`` → ``/private/tmp`` on macOS).
    """
    parts = pat.split(os.sep)
    cut = 0
    for i, part in enumerate(parts):
        if any(c in _GLOB_META for c in part):
            break
        cut = i + 1
    if cut == 0:
        return pat
    literal_parts = parts[:cut]
    prefix = os.sep.join(literal_parts)
    if not prefix:
        return pat
    try:
        resolved = os.path.realpath(prefix, strict=True)
        return _join_pattern(resolved, parts[cut:])
    except (FileNotFoundError, OSError):
        for i in range(len(literal_parts) - 1, 0, -1):
            ancestor = os.sep.join(literal_parts[:i])
            if not ancestor:
                continue
            try:
                resolved = os.path.realpath(ancestor, strict=True)
                tail = literal_parts[i:] + parts[cut:]
                return _join_pattern(resolved, tail)
            except (FileNotFoundError, OSError):
                continue
    return pat


def _join_pattern(resolved: str, rest: list[str]) -> str:
    return resolved if not rest else resolved + os.sep + os.sep.join(rest)


def _op_string(op: Op) -> str:
    parts: list[str] = []
    if op & Op.READ:
        parts.append("read")
    if op & Op.WRITE:
        parts.append("write")
    if op & Op.EXEC:
        parts.append("exec")
    return "|".join(parts) if parts else "none"

File: sdk/ts/src/scope.ts
/**
 * @packageDocumentation
 *
 * Filesystem path policy guardrails — TypeScript port of `hop.top/kit/go/core/scope`.
 *
 * Mirrors the Go API surface (`New`, `Default`, `Allow`, `Deny`, `Check`,
 * `Enforce`, `SetMode`, `SetDefault`, `Snapshot`, `SecretPaths`). The default
 * deny list is loaded from `contracts/parity/scope-defaults.json` (the
 * cross-language source of truth).
 *
 * Glob matching uses [`picomatch`](https://github.com/micromatch/picomatch)
 * which supports the subset of bash-style globs (including `**`) that the Go
 * port relies on via `bmatcuk/doublestar/v4`.
 *
 * @example
 * ```ts
 * import { Default, Op, Strict } from '@hop-top/kit/scope';
 *
 * const dec = Default().check('/Users/me/.ssh/id_rsa', Op.Read);
 * // -> Decision.Denied
 * ```
 */

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import picomatch from 'picomatch';

import scopeDefaultsRaw from './scope-defaults.json';

// ─── Types ───────────────────────────────────────────────────────────────────

/** Mode controls how `Enforce` reacts to a `Denied` decision. */
export enum Mode {
  /** Strict denies → throws `ErrDenied`. The default. */
  Strict = 'strict',
  /** Warn denies → logs to stderr and returns. */
  Warn = 'warn',
  /** Prompt denies → invokes the prompt callback. */
  Prompt = 'prompt',
}

/** Op is a bitset of filesystem operations. */
export enum Op {
  Read = 1 << 0,
  Write = 1 << 1,
  Exec = 1 << 2,
}

/** Decision is the outcome of `Check`. */
export enum Decision {
  Unknown = 'unknown',
  Allowed = 'allowed',
  Denied = 'denied',
}

/** A single rule: glob patterns + ops bitset + allow/deny verdict. */
export interface Rule {
  patterns: string[];
  ops: Op;
  allow: boolean;
}

/** Prompt callback used in `Mode.Prompt`. Return `true` to allow this call. */
export type PromptFunc = (path: string, op: Op) => boolean;

/** Error thrown by `Enforce` when a path is denied. */
export class ErrDenied extends Error {
  constructor(p: string, op: Op) {
    super(`scope: path denied: ${p} (op=${opString(op)})`);
    this.name = 'ErrDenied';
  }
}

// ─── Defaults JSON shape ─────────────────────────────────────────────────────

interface ScopeDefaults {
  version: number;
  deny: Record<string, string[]>;
}

const defaults: ScopeDefaults = scopeDefaultsRaw as unknown as ScopeDefaults;

// ─── Policy ──────────────────────────────────────────────────────────────────

/**
 * Policy holds an ordered list of rules, a Mode, and an optional prompt
 * callback. Construct via `New()` or `Default()`.
 */
export class Policy {
  private rules: Rule[] = [];
  private mode: Mode = Mode.Strict;
  private prompt?: PromptFunc;

  /** Register an allow rule covering Read|Write|Exec. */
  allow(...patterns: string[]): this {
    return this.allowOp(Op.Read | Op.Write | Op.Exec, ...patterns);
  }

  /** Register an allow rule for the given operations. */
  allowOp(op: Op, ...patterns: string[]): this {
    if (patterns.length === 0) return this;
    this.rules.push({ patterns: [...patterns], ops: op, allow: true });
    return this;
  }

  /** Register a deny rule covering Read|Write|Exec. */
  deny(...patterns: string[]): this {
    return this.denyOp(Op.Read | Op.Write | Op.Exec, ...patterns);
  }

  /** Register a deny rule for the given operations. */
  denyOp(op: Op, ...patterns: string[]): this {
    if (patterns.length === 0) return this;
    this.rules.push({ patterns: [...patterns], ops: op, allow: false });
    return this;
  }

  /** Set the enforcement mode. */
  setMode(m: Mode): this {
    this.mode = m;
    return this;
  }

  /** Get the current enforcement mode. */
  getMode(): Mode {
    return this.mode;
  }

  /** Set the prompt callback used in `Mode.Prompt`. */
  setPromptFunc(fn: PromptFunc): this {
    this.prompt = fn;
    return this;
  }

  /** Defensive copy of the policy's rules in registration order. */
  getRules(): Rule[] {
    return this.rules.map((r) => ({ patterns: [...r.patterns], ops: r.ops, allow: r.allow }));
  }

  /**
   * Evaluate `(p, op)` against the policy. Pure — no prompt, no mutation.
   * Resolves symlinks before matching to defeat symlink escapes; on ENOENT
   * the cleaned path is used so "intent to write" still matches deny rules.
   */
  check(p: string, op: Op): Decision {
    const resolved = resolvePath(p);
    let sawAllow = false;
    for (const r of this.rules) {
      if ((r.ops & op) === 0) continue;
      if (!matchAny(r.patterns, resolved)) continue;
      if (!r.allow) return Decision.Denied;
      sawAllow = true;
    }
    return sawAllow ? Decision.Allowed : Decision.Unknown;
  }

  /**
   * Enforce calls `check` and translates the decision according to mode:
   * Strict + (Denied|Unknown) throws; Warn + Denied logs; Prompt + Denied
   * invokes the prompt callback.
   */
  enforce(p: string, op: Op): void {
    const dec = this.check(p, op);
    if (dec === Decision.Allowed) return;
    if (dec === Decision.Unknown && this.mode !== Mode.Strict) return;
    this.handleDeny(p, op);
  }

  private handleDeny(p: string, op: Op): void {
    switch (this.mode) {
      case Mode.Warn:
        console.warn(`scope: path denied (warn mode, allowing): ${p} op=${opString(op)}`);
        return;
      case Mode.Prompt:
        if (this.prompt && this.prompt(p, op)) return;
        throw new ErrDenied(p, op);
      default: // Strict
        throw new ErrDenied(p, op);
    }
  }

  /** Deep copy. Useful for per-test isolation paired with `setDefault`. */
  snapshot(): Policy {
    const cp = new Policy();
    cp.mode = this.mode;
    cp.prompt = this.prompt;
    cp.rules = this.rules.map((r) => ({ patterns: [...r.patterns], ops: r.ops, allow: r.allow }));
    return cp;
  }
}

// ─── Singleton + factory ─────────────────────────────────────────────────────

let defaultP: Policy | null = null;

/** Create an empty `Policy` in `Mode.Strict`. */
export function New(): Policy {
  return new Policy();
}

/**
 * Returns the package-level singleton policy used by primitives that don't
 * accept an explicit policy. Lazy-initialised on first call. The singleton is
 * pre-seeded with `SecretPaths()` as a deny-all-ops rule.
 */
export function Default(): Policy {
  if (!defaultP) {
    defaultP = New().deny(...SecretPaths());
  }
  return defaultP;
}

/**
 * Swap the package-level singleton with `p`. Returns a restore function that
 * puts the previous policy back. Intended for tests:
 *
 * ```ts
 * const restore = setDefault(New().allow('~/tmp/**'));
 * try { ... } finally { restore(); }
 * ```
 */
export function setDefault(p: Policy): () => void {
  const prev = Default();
  defaultP = p;
  return () => {
    defaultP = prev;
  };
}

// ─── Default deny list ───────────────────────────────────────────────────────

/**
 * Returns the default deny pattern set: common patterns plus the patterns
 * specific to the current platform. Patterns are ready to feed into
 * `Policy.deny`. `~` and Windows env macros are resolved at match time.
 */
export function SecretPaths(): string[] {
  const common = defaults.deny['common'] ?? [];
  const platName = platformKey();
  const platSpecific = defaults.deny[platName] ?? [];
  return [...common, ...platSpecific].map(expandWindowsEnv);
}

/** Map Node.js `process.platform` to the keys used in scope-defaults.json. */
function platformKey(): string {
  switch (process.platform) {
    case 'darwin':
      return 'darwin';
    case 'win32':
      return 'windows';
    default:
      return 'linux';
  }
}

/**
 * Expand `%APPDATA%`, `%LOCALAPPDATA%`, `%USERPROFILE%` in `p` using the
 * matching env vars when present. On non-Windows hosts the expansion is
 * best-effort: when the env var is empty the macro is left in place and the
 * pattern simply won't match anything.
 */
function expandWindowsEnv(p: string): string {
  for (const key of ['APPDATA', 'LOCALAPPDATA', 'USERPROFILE']) {
    const token = `%${key}%`;
    if (!p.includes(token)) continue;
    const v = process.env[key];
    if (!v) continue;
    p = p.split(token).join(v);
  }
  return p;
}

// ─── Path resolution + match ─────────────────────────────────────────────────

/** Expand a leading `~` to the user's home dir, then `path.normalize`. */
function expandHome(s: string): string {
  if (s === '~' || s.startsWith('~/')) {
    let home = os.homedir();
    try {
      home = fs.realpathSync(home);
    } catch {
      // best effort; keep raw home
    }
    return s === '~' ? home : path.join(home, s.slice(2));
  }
  return s;
}

/**
 * Resolve `s`: expand `~`, normalize, and follow symlinks. On ENOENT walks
 * up to the deepest existing ancestor and re-attaches the missing tail.
 */
function resolvePath(s: string): string {
  const expanded = expandHome(s);
  const cleaned = path.normalize(expanded);
  try {
    return fs.realpathSync(cleaned);
  } catch (err: unknown) {
    if (!isENOENT(err)) return cleaned;
  }
  // Walk up to the deepest existing ancestor.
  let dir = path.dirname(cleaned);
  let tail = path.basename(cleaned);
  while (dir && dir !== path.parse(dir).root) {
    try {
      const resolved = fs.realpathSync(dir);
      return path.join(resolved, tail);
    } catch (err: unknown) {
      if (!isENOENT(err)) return cleaned;
    }
    tail = path.join(path.basename(dir), tail);
    dir = path.dirname(dir);
  }
  return cleaned;
}

function isENOENT(err: unknown): boolean {
  return typeof err === 'object' && err !== null && (err as { code?: string }).code === 'ENOENT';
}

/** Returns true if any of the patterns matches `abs` (already resolved). */
function matchAny(patterns: string[], abs: string): boolean {
  for (const p of patterns) {
    const expanded = expandHome(p);
    const canon = canonicalisePattern(path.normalize(expanded));
    if (picomatch.isMatch(abs, canon, { dot: true })) return true;
  }
  return false;
}

/**
 * Resolve the longest leading literal directory prefix of `pat` through
 * realpath so the pattern keeps matching after the input path is canonicalised.
 */
function canonicalisePattern(pat: string): string {
  const sep = path.sep;
  const parts = pat.split(sep);
  let cut = 0;
  for (let i = 0; i < parts.length; i++) {
    if (containsGlobMeta(parts[i] ?? '')) break;
    cut = i + 1;
  }
  if (cut === 0) return pat;
  const literalParts = parts.slice(0, cut);
  const prefix = literalParts.join(sep);
  if (!prefix) return pat;
  try {
    const resolved = fs.realpathSync(prefix);
    return joinPattern(resolved, parts.slice(cut));
  } catch {
    // Walk up to the deepest existing ancestor.
    for (let i = literalParts.length - 1; i >= 1; i--) {
      const ancestor = literalParts.slice(0, i).join(sep);
      if (!ancestor) continue;
      try {
        const resolved = fs.realpathSync(ancestor);
        const tail = [...literalParts.slice(i), ...parts.slice(cut)];
        return joinPattern(resolved, tail);
      } catch {
        // continue
      }
    }
  }
  return pat;
}

function joinPattern(resolved: string, rest: string[]): string {
  return rest.length === 0 ? resolved : resolved + path.sep + rest.join(path.sep);
}

function containsGlobMeta(s: string): boolean {
  return /[*?[\]{}]/.test(s);
}

function opString(op: Op): string {
  const parts: string[] = [];
  if (op & Op.Read) parts.push('read');
  if (op & Op.Write) parts.push('write');
  if (op & Op.Exec) parts.push('exec');
  return parts.length === 0 ? 'none' : parts.join('|');
}
